test/e2e/upg_e2e.go

```diff
@@ -425,6 +428,56 @@ func describePDRReplacement(f *framework.Framework) {
 	})
 }
 
+var _ = ginkgo.Describe("Binapi", func() {
+	ginkgo.Context("policy based routing", func() {
+		f := framework.NewDefaultFramework(framework.UPGModeTDF, framework.UPGIPModeV4)
+		ginkgo.It("Check binapi functions", func() {
+			policy := &upf.UpfPolicyAddDel{}
+			policy.Action = 1
+			policy.Identifier = "qwerty"
+			policy.NPaths = 1
+			rpath := fib_types.FibPath{}
+			nhip, err := ip_types.ParseAddress("144.0.0.2")
+			gomega.Expect(err).To(gomega.BeNil())
+			rpath.Nh.Address.SetIP4(nhip.Un.GetIP4())
+			rpath.SwIfIndex = 3
+			rpath.Proto = fib_types.FIB_API_PATH_NH_PROTO_IP4
+			rpath.Flags = 0
+			policy.Paths = append(policy.Paths, rpath)
+
+			policyReply := &upf.UpfPolicyAddDelReply{}
+			err = f.VPP.ApiChannel.SendRequest(policy).ReceiveReply(policyReply)
+			gomega.Expect(err).To(gomega.BeNil())
+
+			reqCtx := f.VPP.ApiChannel.SendMultiRequest(&upf.UpfPolicyDump{})
+
+			for {
+				msg := &upf.UpfPolicyDetails{}
+				stop, err := reqCtx.ReceiveReply(msg)
+				gomega.Expect(err).To(gomega.BeNil())
+				if stop {
+					break
+				}
+				gomega.Expect(msg.Identifier).To(gomega.BeEquivalentTo(policy.Identifier))
+				gomega.Expect(msg.NPaths).To(gomega.BeEquivalentTo(1))
+				for i := 0; i < int(msg.NPaths); i++ {
+					gomega.Expect(msg.Paths[i].SwIfIndex).To(gomega.BeEquivalentTo(policy.Paths[i].SwIfIndex))
+					gomega.Expect(msg.Paths[i].Nh.Address.GetIP4().String()).To(gomega.BeEquivalentTo(policy.Paths[i].Nh.Address.GetIP4().String()))
+				}
+			}
+
+			policy.Action = 0
+			policyReply = &upf.UpfPolicyAddDelReply{}
+			err = f.VPP.ApiChannel.SendRequest(policy).ReceiveReply(policyReply)
+			gomega.Expect(err).To(gomega.BeNil())
+
+			msg := &upf.UpfPolicyDetails{}
+			_, err = reqCtx.ReceiveReply(msg)
+			gomega.Expect(err).NotTo(gomega.BeNil())
+		})
```

Comment:
This does not verify that the routing policy does actually work, which would probably involve adding something that can be used to test routing to the SGi side. We probably should have TODO here and an issue for that task; otherwise the functionality is effectively untested

Reply:
I've added an UPG issue #189 that should be taken into account